File: java/scf-core/src/main/java/org/mydotey/scf/impl/AbstractConfigurationSource.java
package org.mydotey.scf.impl;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.function.Consumer;

import org.mydotey.scf.ConfigurationSource;
import org.mydotey.scf.ConfigurationSourceConfig;
import org.mydotey.scf.PropertyConfig;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * @author koqizhao
 *
 * May 16, 2018
 */
public abstract class AbstractConfigurationSource implements ConfigurationSource {

    private static final Logger LOGGER = LoggerFactory.getLogger(AbstractConfigurationSource.class);

    private ConfigurationSourceConfig _config;

    private boolean _dynamic;
    private volatile List<Consumer<ConfigurationSource>> _changeListeners;

    public AbstractConfigurationSource(ConfigurationSourceConfig config) {
        Objects.requireNonNull(config, "config is null");

        _config = config;
    }

    @Override
    public ConfigurationSourceConfig getConfig() {
        return _config;
    }

    @Override
    public <K, V> V getPropertyValue(PropertyConfig<K, V> propertyConfig) {
        try {
            return doGetPropertyValue(propertyConfig);
        } catch (Exception e) {
            LOGGER.error("source getPropertyValue failed to run", e);
            return null;
        }
    }

    protected abstract <K, V> V doGetPropertyValue(PropertyConfig<K, V> propertyConfig);

    @Override
    public boolean isDynamic() {
        return _dynamic;
    }

    protected void setDynamic(boolean dynamic) {
        _dynamic = dynamic;
    }

    @Override
    public void addChangeListener(Consumer<ConfigurationSource> changeListener) {
        Objects.requireNonNull("changeListener", "changeListener is null");

        if (!_dynamic)
            return;

        synchronized (this) {
            if (_changeListeners == null)
                _changeListeners = new ArrayList<>();

            _changeListeners.add(changeListener);
        }
    }

    protected void raiseChangeEvent() {
        if (!_dynamic)
            return;

        synchronized (this) {
            if (_changeListeners == null)
                return;

            _changeListeners.forEach(l -> {
                try {
                    l.accept(AbstractConfigurationSource.this);
                } catch (Exception e) {
                    LOGGER.error("source change listener failed to run", e);
                }
            });
        }
    }

}

File: dotnet/src/MyDotey.SCF/AbstractConfigurationSource.cs
using System;
using System.Collections;
using System.Collections.Generic;

using NLog;
using MyDotey.SCF.Type;

namespace MyDotey.SCF
{
    /**
     * @author koqizhao
     *
     * May 16, 2018
     */
    public abstract class AbstractConfigurationSource<C> : IConfigurationSource
        where C : ConfigurationSourceConfig
    {
        private static readonly Logger Logger = LogManager.GetCurrentClassLogger(typeof(AbstractConfigurationSource<>));

        private static readonly ICollection<ITypeConverter> EmptyValueConverters = new List<ITypeConverter>();

        private C _config;

        private volatile List<EventHandler<IConfigurationSourceChangeEvent>> _changeListeners;

        public AbstractConfigurationSource(C config)
        {
            if (config == null)
                throw new ArgumentNullException("config is null");

            _config = config;
        }

        ConfigurationSourceConfig IConfigurationSource.Config { get { return Config; } }

        public virtual C Config { get { return _config; } }

        public virtual event EventHandler<IConfigurationSourceChangeEvent> OnChange
        {
            add
            {
                if (value == null)
                    throw new ArgumentNullException("changeListener is null");

                lock (this)
                {
                    if (_changeListeners == null)
                        _changeListeners = new List<EventHandler<IConfigurationSourceChangeEvent>>();

                    _changeListeners.Add(value);
                }
            }

            remove
            {
                throw new NotSupportedException("remove is not supported");
            }
        }

        protected virtual void RaiseChangeEvent()
        {
            lock (this)
            {
                if (_changeListeners == null)
                    return;

                _changeListeners.ForEach(l =>
                {
                    try
                    {
                        l(this, new DefaultConfigurationSourceChangeEvent(this));
                    }
                    catch (Exception e)
                    {
                        Logger.Error(e, "source change listener failed to run");
                    }
                });
            }
        }

        public virtual V GetPropertyValue<K, V>(PropertyConfig<K, V> propertyConfig)
        {
            object value = GetPropertyValue(propertyConfig.Key);
            return Convert(propertyConfig, value);
        }

        protected abstract object GetPropertyValue(object key);

        protected virtual V Convert<K, V>(PropertyConfig<K, V> propertyConfig, object value)
        {
            if (IsDefault<V>(value))
                return default(V);

            ICollection<ITypeConverter> valueConverters = propertyConfig.ValueConverters ?? EmptyValueConverters;
            foreach (ITypeConverter typeConverter in valueConverters)
            {
                if (typeConverter.SourceType.IsAssignableFrom(value.GetType())
                        && typeof(V).IsAssignableFrom(typeConverter.TargetType))
                {
                    V v = (V)typeConverter.Convert(value);
                    if (v != null)
                        return v;
                }
            }

            return value is V ? (V)value : default(V);
        }

        protected virtual bool IsDefault<V>(object value)
        {
            if (Object.Equals(value, default(V)))
                return true;

            if (value is string)
                return string.IsNullOrWhiteSpace((string)value);

            if (value is ICollection)
                return ((ICollection)value).Count == 0;

            if (value is Array)
                return ((Array)value).Length == 0;

            return false;
        }

        public override string ToString()
        {
            return string.Format("{0} {{ config: {1}, changeListeners: [ {2} ] }}", GetType().Name, Config,
                    _changeListeners == null ? null : string.Join(", ", _changeListeners));
        }
    }
}